{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import python libraries  \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import tensorflow as tf\n",
    "import sklearn\n",
    "import sys\n",
    "import os\n",
    "import time\n",
    "import scipy\n",
    "import pickle\n",
    "from importlib import reload\n",
    "from sklearn.metrics import r2_score, mean_squared_error\n",
    "from scipy.stats import pearsonr\n",
    "from tensorflow.keras import layers\n",
    "codebase_path = '/data/home/wpw035/Codebase'\n",
    "#sys.path.insert(0, codebase_path) #add path to my codebase models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from DRP_nb import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'DRP_nb.cross_val' from '/data/home/wpw035/Drug_response_prediction/DRP_exmaple_msc/drug_response_prediction_example/DRP_nb/cross_val.py'>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#my libary imports, code in DRP_nb file\n",
    "from DRP_nb import data_imports, splitting, hp_opt, utils, mean_model, cross_val\n",
    "reload(data_imports)\n",
    "reload(splitting)\n",
    "reload(utils)\n",
    "reload(cross_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Imports\n",
    "The data loading and cleaning are done within the data_imports module within my DRP_nb library. The data loading and cleaning are done within the data_imports module within my DRP_nb library. Take a look at the data_imports file within DRP_nb to see all the steps. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/home/wpw035/.conda/envs/tfGPUforge/lib/python3.9/site-packages/pandas/util/_decorators.py:311: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support skipfooter; you can avoid this warning by specifying engine='python'.\n",
      "  return func(*args, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "% of missing prot values 0.386335609896865\n",
      "num non overlapping prot and target cls: 10\n",
      "num non overlapping cls: 930\n"
     ]
    }
   ],
   "source": [
    "#omics and target imports \n",
    "prot, rna, phospho_ls, ic50_df1, one_hot_cls, one_hot_drugs = data_imports.read_inputs_phos_ect()\n",
    "_all_cls = prot.index\n",
    "_all_drugs = one_hot_drugs.columns\n",
    "\n",
    "\n",
    "#load landmark genes that will be used for feature selection\n",
    "landmark_genes = pd.read_csv(\n",
    "    f'{codebase_path}/downloaded_data_small/landmark_genes_LINCS.txt',sep='\\t')\n",
    "landmark_genes.index = landmark_genes['Symbol']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature selection \n",
    "Feature selection using the 1K landmark genes\n",
    "\n",
    "will foucs on RNA-seq data\n",
    "Only keep an feature from the rna df if it is an landmark gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#finds landmarks that are also in rna df\n",
    "overlapping_landmarks = [g for g in landmark_genes.index if g in rna.columns]\n",
    "rna = rna[overlapping_landmarks]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(38, 908)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GENE_SYMBOLS</th>\n",
       "      <th>INSIG1</th>\n",
       "      <th>FOXO3</th>\n",
       "      <th>CDH3</th>\n",
       "      <th>ORC1</th>\n",
       "      <th>ITGAE</th>\n",
       "      <th>CDK7</th>\n",
       "      <th>FOS</th>\n",
       "      <th>CDK6</th>\n",
       "      <th>CDK4</th>\n",
       "      <th>CDK2</th>\n",
       "      <th>...</th>\n",
       "      <th>FHL2</th>\n",
       "      <th>IL13RA1</th>\n",
       "      <th>FKBP4</th>\n",
       "      <th>ABHD4</th>\n",
       "      <th>HIST1H2BK</th>\n",
       "      <th>NUCB2</th>\n",
       "      <th>NUP88</th>\n",
       "      <th>NVL</th>\n",
       "      <th>ILK</th>\n",
       "      <th>INPP1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>COLO-680N</th>\n",
       "      <td>8.154784</td>\n",
       "      <td>9.236981</td>\n",
       "      <td>9.799767</td>\n",
       "      <td>5.103765</td>\n",
       "      <td>10.757260</td>\n",
       "      <td>9.690477</td>\n",
       "      <td>9.034680</td>\n",
       "      <td>6.728485</td>\n",
       "      <td>8.812277</td>\n",
       "      <td>6.160667</td>\n",
       "      <td>...</td>\n",
       "      <td>8.719871</td>\n",
       "      <td>6.707274</td>\n",
       "      <td>7.683773</td>\n",
       "      <td>3.787795</td>\n",
       "      <td>9.465148</td>\n",
       "      <td>7.516577</td>\n",
       "      <td>7.447703</td>\n",
       "      <td>6.983274</td>\n",
       "      <td>6.795911</td>\n",
       "      <td>9.728236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDM-1</th>\n",
       "      <td>8.055311</td>\n",
       "      <td>8.088824</td>\n",
       "      <td>3.379974</td>\n",
       "      <td>6.465721</td>\n",
       "      <td>9.480014</td>\n",
       "      <td>8.702067</td>\n",
       "      <td>3.157377</td>\n",
       "      <td>9.285379</td>\n",
       "      <td>8.979623</td>\n",
       "      <td>6.235116</td>\n",
       "      <td>...</td>\n",
       "      <td>6.329245</td>\n",
       "      <td>4.010940</td>\n",
       "      <td>9.105476</td>\n",
       "      <td>3.394924</td>\n",
       "      <td>8.919325</td>\n",
       "      <td>8.434682</td>\n",
       "      <td>7.330557</td>\n",
       "      <td>6.003414</td>\n",
       "      <td>6.626827</td>\n",
       "      <td>8.023713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HEL</th>\n",
       "      <td>7.678645</td>\n",
       "      <td>8.126148</td>\n",
       "      <td>3.383949</td>\n",
       "      <td>7.070015</td>\n",
       "      <td>8.819721</td>\n",
       "      <td>8.586132</td>\n",
       "      <td>6.420190</td>\n",
       "      <td>8.927892</td>\n",
       "      <td>9.720554</td>\n",
       "      <td>6.398864</td>\n",
       "      <td>...</td>\n",
       "      <td>9.807979</td>\n",
       "      <td>3.422834</td>\n",
       "      <td>7.842642</td>\n",
       "      <td>3.879684</td>\n",
       "      <td>11.686403</td>\n",
       "      <td>7.181875</td>\n",
       "      <td>6.733916</td>\n",
       "      <td>6.101298</td>\n",
       "      <td>6.616129</td>\n",
       "      <td>9.701949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HL-60</th>\n",
       "      <td>9.919567</td>\n",
       "      <td>7.989809</td>\n",
       "      <td>3.171409</td>\n",
       "      <td>7.077472</td>\n",
       "      <td>8.133847</td>\n",
       "      <td>8.347510</td>\n",
       "      <td>6.516647</td>\n",
       "      <td>8.619618</td>\n",
       "      <td>9.546260</td>\n",
       "      <td>6.068342</td>\n",
       "      <td>...</td>\n",
       "      <td>3.166125</td>\n",
       "      <td>4.936508</td>\n",
       "      <td>7.779411</td>\n",
       "      <td>3.562882</td>\n",
       "      <td>11.235920</td>\n",
       "      <td>10.315398</td>\n",
       "      <td>6.815571</td>\n",
       "      <td>6.391719</td>\n",
       "      <td>6.212755</td>\n",
       "      <td>5.730575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JHH-2</th>\n",
       "      <td>9.251225</td>\n",
       "      <td>8.227397</td>\n",
       "      <td>3.260621</td>\n",
       "      <td>3.861217</td>\n",
       "      <td>9.561528</td>\n",
       "      <td>8.988581</td>\n",
       "      <td>5.048408</td>\n",
       "      <td>8.471914</td>\n",
       "      <td>8.982768</td>\n",
       "      <td>5.421258</td>\n",
       "      <td>...</td>\n",
       "      <td>9.257869</td>\n",
       "      <td>8.454283</td>\n",
       "      <td>6.429325</td>\n",
       "      <td>5.200694</td>\n",
       "      <td>9.746995</td>\n",
       "      <td>8.678383</td>\n",
       "      <td>6.247309</td>\n",
       "      <td>6.080807</td>\n",
       "      <td>7.977128</td>\n",
       "      <td>8.423243</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 908 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GENE_SYMBOLS    INSIG1     FOXO3      CDH3      ORC1      ITGAE      CDK7  \\\n",
       "COLO-680N     8.154784  9.236981  9.799767  5.103765  10.757260  9.690477   \n",
       "GDM-1         8.055311  8.088824  3.379974  6.465721   9.480014  8.702067   \n",
       "HEL           7.678645  8.126148  3.383949  7.070015   8.819721  8.586132   \n",
       "HL-60         9.919567  7.989809  3.171409  7.077472   8.133847  8.347510   \n",
       "JHH-2         9.251225  8.227397  3.260621  3.861217   9.561528  8.988581   \n",
       "\n",
       "GENE_SYMBOLS       FOS      CDK6      CDK4      CDK2  ...      FHL2   IL13RA1  \\\n",
       "COLO-680N     9.034680  6.728485  8.812277  6.160667  ...  8.719871  6.707274   \n",
       "GDM-1         3.157377  9.285379  8.979623  6.235116  ...  6.329245  4.010940   \n",
       "HEL           6.420190  8.927892  9.720554  6.398864  ...  9.807979  3.422834   \n",
       "HL-60         6.516647  8.619618  9.546260  6.068342  ...  3.166125  4.936508   \n",
       "JHH-2         5.048408  8.471914  8.982768  5.421258  ...  9.257869  8.454283   \n",
       "\n",
       "GENE_SYMBOLS     FKBP4     ABHD4  HIST1H2BK      NUCB2     NUP88       NVL  \\\n",
       "COLO-680N     7.683773  3.787795   9.465148   7.516577  7.447703  6.983274   \n",
       "GDM-1         9.105476  3.394924   8.919325   8.434682  7.330557  6.003414   \n",
       "HEL           7.842642  3.879684  11.686403   7.181875  6.733916  6.101298   \n",
       "HL-60         7.779411  3.562882  11.235920  10.315398  6.815571  6.391719   \n",
       "JHH-2         6.429325  5.200694   9.746995   8.678383  6.247309  6.080807   \n",
       "\n",
       "GENE_SYMBOLS       ILK     INPP1  \n",
       "COLO-680N     6.795911  9.728236  \n",
       "GDM-1         6.626827  8.023713  \n",
       "HEL           6.616129  9.701949  \n",
       "HL-60         6.212755  5.730575  \n",
       "JHH-2         7.977128  8.423243  \n",
       "\n",
       "[5 rows x 908 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(rna.shape)\n",
    "rna.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above shows the rna df has 38 rows and 908 columns. Thus, it gives the expression values for 38 cell lines for 908 landmark genes that were in the original rna-seq dataset. \n",
    "\n",
    "\n",
    "The input we want for drug response prediction is drug cell line pairs because we want to predict the efficacy of a given drug for a given cell. Therefore, the next step is to use the df we have to create inputs for all drug cell line pairs. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phos feature selection\n",
    "for phosphoproteomics data keep the targets of landmark genes that are also landmark genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "186\n",
      "1064\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>col.name</th>\n",
       "      <th>EGFR(T693);</th>\n",
       "      <th>EGFR(S695);</th>\n",
       "      <th>EGFR(S1166);</th>\n",
       "      <th>EGFR(S991);</th>\n",
       "      <th>EGFR(T993);</th>\n",
       "      <th>EGFR(S995);</th>\n",
       "      <th>EGFR(T1032);EGFR(T1046);</th>\n",
       "      <th>EGFR(T1046);</th>\n",
       "      <th>EGFR(S1042);EGFR(S1045);</th>\n",
       "      <th>EGFR(S1045);</th>\n",
       "      <th>...</th>\n",
       "      <th>EML3(S157);</th>\n",
       "      <th>EML3(S156);</th>\n",
       "      <th>EML3(S889);</th>\n",
       "      <th>EML3(S883);</th>\n",
       "      <th>EZH2(T388);</th>\n",
       "      <th>EZH2(T487);</th>\n",
       "      <th>EZH2(S362);</th>\n",
       "      <th>EZH2(T369);</th>\n",
       "      <th>EZH2(S362);EZH2(S363);</th>\n",
       "      <th>EZH2(S366);</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>COLO-680N</th>\n",
       "      <td>0.870410</td>\n",
       "      <td>0.859884</td>\n",
       "      <td>0.487725</td>\n",
       "      <td>0.522972</td>\n",
       "      <td>0.688719</td>\n",
       "      <td>-1.130207</td>\n",
       "      <td>0.948421</td>\n",
       "      <td>0.948421</td>\n",
       "      <td>-1.130207</td>\n",
       "      <td>-1.130207</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.287052</td>\n",
       "      <td>-1.287052</td>\n",
       "      <td>-1.287052</td>\n",
       "      <td>0.627996</td>\n",
       "      <td>-1.287052</td>\n",
       "      <td>1.059854</td>\n",
       "      <td>1.246416</td>\n",
       "      <td>0.640216</td>\n",
       "      <td>1.215685</td>\n",
       "      <td>-1.287052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDM-1</th>\n",
       "      <td>-1.429010</td>\n",
       "      <td>-1.429010</td>\n",
       "      <td>-1.429010</td>\n",
       "      <td>0.430902</td>\n",
       "      <td>0.578810</td>\n",
       "      <td>-1.271041</td>\n",
       "      <td>-0.048920</td>\n",
       "      <td>-0.048920</td>\n",
       "      <td>-1.271041</td>\n",
       "      <td>-1.271041</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.429010</td>\n",
       "      <td>0.754475</td>\n",
       "      <td>0.288527</td>\n",
       "      <td>0.462040</td>\n",
       "      <td>0.185137</td>\n",
       "      <td>1.045595</td>\n",
       "      <td>1.414728</td>\n",
       "      <td>0.925482</td>\n",
       "      <td>1.374856</td>\n",
       "      <td>-1.429010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HEL</th>\n",
       "      <td>-1.613351</td>\n",
       "      <td>-1.613351</td>\n",
       "      <td>-1.613351</td>\n",
       "      <td>0.336451</td>\n",
       "      <td>0.457029</td>\n",
       "      <td>-1.463476</td>\n",
       "      <td>-1.313602</td>\n",
       "      <td>-1.313602</td>\n",
       "      <td>-1.463476</td>\n",
       "      <td>-1.463476</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.613351</td>\n",
       "      <td>-1.613351</td>\n",
       "      <td>0.349405</td>\n",
       "      <td>0.773424</td>\n",
       "      <td>-1.613351</td>\n",
       "      <td>0.970619</td>\n",
       "      <td>1.327014</td>\n",
       "      <td>0.845071</td>\n",
       "      <td>1.271451</td>\n",
       "      <td>0.492025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HL-60</th>\n",
       "      <td>0.426137</td>\n",
       "      <td>-1.428863</td>\n",
       "      <td>-1.428863</td>\n",
       "      <td>-1.206791</td>\n",
       "      <td>0.335992</td>\n",
       "      <td>-1.288751</td>\n",
       "      <td>0.373059</td>\n",
       "      <td>0.373059</td>\n",
       "      <td>0.403272</td>\n",
       "      <td>0.403272</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.428863</td>\n",
       "      <td>-1.428863</td>\n",
       "      <td>0.357729</td>\n",
       "      <td>0.508261</td>\n",
       "      <td>-1.428863</td>\n",
       "      <td>0.727035</td>\n",
       "      <td>1.430922</td>\n",
       "      <td>0.995132</td>\n",
       "      <td>1.396240</td>\n",
       "      <td>0.468569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JHH-2</th>\n",
       "      <td>0.713488</td>\n",
       "      <td>0.820889</td>\n",
       "      <td>0.532149</td>\n",
       "      <td>0.686185</td>\n",
       "      <td>1.006834</td>\n",
       "      <td>-1.076954</td>\n",
       "      <td>0.530054</td>\n",
       "      <td>0.530054</td>\n",
       "      <td>-1.076954</td>\n",
       "      <td>-1.076954</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.224318</td>\n",
       "      <td>-1.224318</td>\n",
       "      <td>0.607585</td>\n",
       "      <td>0.410201</td>\n",
       "      <td>-1.224318</td>\n",
       "      <td>0.906618</td>\n",
       "      <td>1.055518</td>\n",
       "      <td>0.600364</td>\n",
       "      <td>1.031041</td>\n",
       "      <td>-1.224318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JHH-4</th>\n",
       "      <td>0.770449</td>\n",
       "      <td>0.857518</td>\n",
       "      <td>0.400208</td>\n",
       "      <td>0.617482</td>\n",
       "      <td>0.778469</td>\n",
       "      <td>0.272771</td>\n",
       "      <td>0.784972</td>\n",
       "      <td>0.784972</td>\n",
       "      <td>0.338587</td>\n",
       "      <td>0.338587</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.273965</td>\n",
       "      <td>-1.273965</td>\n",
       "      <td>0.362711</td>\n",
       "      <td>0.591988</td>\n",
       "      <td>-1.273965</td>\n",
       "      <td>1.174851</td>\n",
       "      <td>1.317839</td>\n",
       "      <td>0.716005</td>\n",
       "      <td>1.273343</td>\n",
       "      <td>0.322767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KASUMI-1</th>\n",
       "      <td>-1.316937</td>\n",
       "      <td>-1.316937</td>\n",
       "      <td>-1.316937</td>\n",
       "      <td>-1.091034</td>\n",
       "      <td>-1.031879</td>\n",
       "      <td>-1.174408</td>\n",
       "      <td>0.209470</td>\n",
       "      <td>0.209470</td>\n",
       "      <td>0.408179</td>\n",
       "      <td>0.408179</td>\n",
       "      <td>...</td>\n",
       "      <td>0.184882</td>\n",
       "      <td>-1.316937</td>\n",
       "      <td>0.650288</td>\n",
       "      <td>0.609231</td>\n",
       "      <td>0.299954</td>\n",
       "      <td>0.797801</td>\n",
       "      <td>1.375823</td>\n",
       "      <td>1.048650</td>\n",
       "      <td>1.319048</td>\n",
       "      <td>-1.316937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KG-1</th>\n",
       "      <td>-1.570566</td>\n",
       "      <td>-1.570566</td>\n",
       "      <td>-0.100433</td>\n",
       "      <td>0.408062</td>\n",
       "      <td>-1.263924</td>\n",
       "      <td>-1.417245</td>\n",
       "      <td>-1.263924</td>\n",
       "      <td>-1.263924</td>\n",
       "      <td>-1.417245</td>\n",
       "      <td>-1.417245</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.570566</td>\n",
       "      <td>0.770972</td>\n",
       "      <td>0.431089</td>\n",
       "      <td>0.649533</td>\n",
       "      <td>-1.570566</td>\n",
       "      <td>0.890752</td>\n",
       "      <td>1.019903</td>\n",
       "      <td>0.291344</td>\n",
       "      <td>0.984688</td>\n",
       "      <td>0.126431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMOE-2</th>\n",
       "      <td>-0.204030</td>\n",
       "      <td>-1.659307</td>\n",
       "      <td>0.168107</td>\n",
       "      <td>0.304357</td>\n",
       "      <td>0.350523</td>\n",
       "      <td>-1.502197</td>\n",
       "      <td>-0.092908</td>\n",
       "      <td>-0.092908</td>\n",
       "      <td>-1.502197</td>\n",
       "      <td>-1.502197</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.144783</td>\n",
       "      <td>0.784860</td>\n",
       "      <td>0.222212</td>\n",
       "      <td>0.502510</td>\n",
       "      <td>0.156714</td>\n",
       "      <td>1.072007</td>\n",
       "      <td>1.310393</td>\n",
       "      <td>0.838632</td>\n",
       "      <td>1.248473</td>\n",
       "      <td>0.521745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KYSE-140</th>\n",
       "      <td>-1.231059</td>\n",
       "      <td>-1.231059</td>\n",
       "      <td>0.144506</td>\n",
       "      <td>0.192449</td>\n",
       "      <td>-0.913721</td>\n",
       "      <td>-1.072390</td>\n",
       "      <td>0.553152</td>\n",
       "      <td>0.553152</td>\n",
       "      <td>-1.072390</td>\n",
       "      <td>-1.072390</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.231059</td>\n",
       "      <td>-1.231059</td>\n",
       "      <td>-1.231059</td>\n",
       "      <td>0.503357</td>\n",
       "      <td>0.507232</td>\n",
       "      <td>1.240663</td>\n",
       "      <td>1.300372</td>\n",
       "      <td>0.752033</td>\n",
       "      <td>1.251735</td>\n",
       "      <td>-1.231059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KYSE-150</th>\n",
       "      <td>-1.259273</td>\n",
       "      <td>0.570344</td>\n",
       "      <td>-1.259273</td>\n",
       "      <td>0.555773</td>\n",
       "      <td>0.675030</td>\n",
       "      <td>-1.104834</td>\n",
       "      <td>0.499938</td>\n",
       "      <td>0.499938</td>\n",
       "      <td>-1.104834</td>\n",
       "      <td>-1.104834</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.259273</td>\n",
       "      <td>-1.259273</td>\n",
       "      <td>0.422692</td>\n",
       "      <td>0.653158</td>\n",
       "      <td>0.260699</td>\n",
       "      <td>1.073071</td>\n",
       "      <td>1.249470</td>\n",
       "      <td>0.706562</td>\n",
       "      <td>1.217383</td>\n",
       "      <td>-1.259273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KYSE-410</th>\n",
       "      <td>-1.307672</td>\n",
       "      <td>0.314822</td>\n",
       "      <td>0.757654</td>\n",
       "      <td>0.457031</td>\n",
       "      <td>-0.985960</td>\n",
       "      <td>-1.146816</td>\n",
       "      <td>-0.985960</td>\n",
       "      <td>-0.985960</td>\n",
       "      <td>-1.146816</td>\n",
       "      <td>-1.146816</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.307672</td>\n",
       "      <td>-1.307672</td>\n",
       "      <td>0.215488</td>\n",
       "      <td>0.671119</td>\n",
       "      <td>0.207402</td>\n",
       "      <td>1.156456</td>\n",
       "      <td>1.230030</td>\n",
       "      <td>0.547308</td>\n",
       "      <td>1.198218</td>\n",
       "      <td>-1.307672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KYSE-450</th>\n",
       "      <td>0.857134</td>\n",
       "      <td>0.980034</td>\n",
       "      <td>0.208004</td>\n",
       "      <td>0.416529</td>\n",
       "      <td>0.349901</td>\n",
       "      <td>-1.081212</td>\n",
       "      <td>0.300771</td>\n",
       "      <td>0.300771</td>\n",
       "      <td>-1.081212</td>\n",
       "      <td>-1.081212</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.244082</td>\n",
       "      <td>-1.244082</td>\n",
       "      <td>0.593143</td>\n",
       "      <td>0.915016</td>\n",
       "      <td>-1.244082</td>\n",
       "      <td>1.160151</td>\n",
       "      <td>1.153924</td>\n",
       "      <td>0.406590</td>\n",
       "      <td>1.100142</td>\n",
       "      <td>-1.244082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KYSE-510</th>\n",
       "      <td>0.924416</td>\n",
       "      <td>-0.846311</td>\n",
       "      <td>-0.846311</td>\n",
       "      <td>-0.608318</td>\n",
       "      <td>-0.545998</td>\n",
       "      <td>-0.696154</td>\n",
       "      <td>-0.545998</td>\n",
       "      <td>-0.545998</td>\n",
       "      <td>-0.696154</td>\n",
       "      <td>-0.696154</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.846311</td>\n",
       "      <td>-0.846311</td>\n",
       "      <td>-0.846311</td>\n",
       "      <td>-0.846311</td>\n",
       "      <td>-0.846311</td>\n",
       "      <td>1.303020</td>\n",
       "      <td>1.618310</td>\n",
       "      <td>1.151798</td>\n",
       "      <td>1.591581</td>\n",
       "      <td>-0.846311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KYSE-520</th>\n",
       "      <td>0.364092</td>\n",
       "      <td>0.691327</td>\n",
       "      <td>0.587081</td>\n",
       "      <td>0.421782</td>\n",
       "      <td>-0.863160</td>\n",
       "      <td>-1.025986</td>\n",
       "      <td>0.284347</td>\n",
       "      <td>0.284347</td>\n",
       "      <td>-1.025986</td>\n",
       "      <td>-1.025986</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.188813</td>\n",
       "      <td>-1.188813</td>\n",
       "      <td>0.239823</td>\n",
       "      <td>0.769691</td>\n",
       "      <td>-1.188813</td>\n",
       "      <td>1.326737</td>\n",
       "      <td>1.276652</td>\n",
       "      <td>0.532564</td>\n",
       "      <td>1.193026</td>\n",
       "      <td>0.341449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KYSE-70</th>\n",
       "      <td>-0.051558</td>\n",
       "      <td>0.269311</td>\n",
       "      <td>-1.335412</td>\n",
       "      <td>0.168138</td>\n",
       "      <td>0.513250</td>\n",
       "      <td>-1.175207</td>\n",
       "      <td>-0.037395</td>\n",
       "      <td>-0.037395</td>\n",
       "      <td>-1.175207</td>\n",
       "      <td>-1.175207</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.335412</td>\n",
       "      <td>-1.335412</td>\n",
       "      <td>-1.335412</td>\n",
       "      <td>0.441229</td>\n",
       "      <td>-0.389097</td>\n",
       "      <td>1.025841</td>\n",
       "      <td>1.172570</td>\n",
       "      <td>0.575567</td>\n",
       "      <td>1.134710</td>\n",
       "      <td>-1.335412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ME-1</th>\n",
       "      <td>-1.543695</td>\n",
       "      <td>-1.543695</td>\n",
       "      <td>-0.007512</td>\n",
       "      <td>-1.319835</td>\n",
       "      <td>0.401082</td>\n",
       "      <td>-1.402455</td>\n",
       "      <td>-1.261215</td>\n",
       "      <td>-1.261215</td>\n",
       "      <td>-1.402455</td>\n",
       "      <td>-1.402455</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.543695</td>\n",
       "      <td>0.142679</td>\n",
       "      <td>0.493540</td>\n",
       "      <td>0.499472</td>\n",
       "      <td>0.065131</td>\n",
       "      <td>1.084595</td>\n",
       "      <td>1.371463</td>\n",
       "      <td>0.951912</td>\n",
       "      <td>1.330122</td>\n",
       "      <td>0.493238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ML-2</th>\n",
       "      <td>-1.598369</td>\n",
       "      <td>-1.598369</td>\n",
       "      <td>-1.598369</td>\n",
       "      <td>-1.365840</td>\n",
       "      <td>-0.040885</td>\n",
       "      <td>-1.451659</td>\n",
       "      <td>0.353061</td>\n",
       "      <td>0.353061</td>\n",
       "      <td>0.030200</td>\n",
       "      <td>0.030200</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.598369</td>\n",
       "      <td>0.385900</td>\n",
       "      <td>0.297178</td>\n",
       "      <td>0.448227</td>\n",
       "      <td>-0.199134</td>\n",
       "      <td>0.925116</td>\n",
       "      <td>1.231671</td>\n",
       "      <td>1.023237</td>\n",
       "      <td>1.154168</td>\n",
       "      <td>-1.598369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOLM-13</th>\n",
       "      <td>0.256812</td>\n",
       "      <td>0.061433</td>\n",
       "      <td>-1.511135</td>\n",
       "      <td>0.282191</td>\n",
       "      <td>0.381162</td>\n",
       "      <td>-1.360741</td>\n",
       "      <td>-1.210347</td>\n",
       "      <td>-1.210347</td>\n",
       "      <td>-1.360741</td>\n",
       "      <td>-1.360741</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.096317</td>\n",
       "      <td>0.889257</td>\n",
       "      <td>0.453363</td>\n",
       "      <td>0.565244</td>\n",
       "      <td>0.089339</td>\n",
       "      <td>0.344541</td>\n",
       "      <td>1.249249</td>\n",
       "      <td>0.769004</td>\n",
       "      <td>1.161006</td>\n",
       "      <td>0.461591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MONO-MAC-6</th>\n",
       "      <td>-1.652866</td>\n",
       "      <td>-1.652866</td>\n",
       "      <td>-1.652866</td>\n",
       "      <td>-1.433589</td>\n",
       "      <td>0.079342</td>\n",
       "      <td>-1.514517</td>\n",
       "      <td>-1.376169</td>\n",
       "      <td>-1.376169</td>\n",
       "      <td>-1.514517</td>\n",
       "      <td>-1.514517</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.652866</td>\n",
       "      <td>-0.845395</td>\n",
       "      <td>0.517394</td>\n",
       "      <td>0.634201</td>\n",
       "      <td>-0.249320</td>\n",
       "      <td>0.439550</td>\n",
       "      <td>1.075657</td>\n",
       "      <td>0.847440</td>\n",
       "      <td>0.976697</td>\n",
       "      <td>0.486960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MV-4-11</th>\n",
       "      <td>-0.429467</td>\n",
       "      <td>-1.530044</td>\n",
       "      <td>-1.530044</td>\n",
       "      <td>0.276384</td>\n",
       "      <td>0.322332</td>\n",
       "      <td>-1.371465</td>\n",
       "      <td>0.205197</td>\n",
       "      <td>0.205197</td>\n",
       "      <td>-1.371465</td>\n",
       "      <td>-1.371465</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.530044</td>\n",
       "      <td>-1.530044</td>\n",
       "      <td>0.577106</td>\n",
       "      <td>0.632136</td>\n",
       "      <td>0.255899</td>\n",
       "      <td>1.101878</td>\n",
       "      <td>1.252620</td>\n",
       "      <td>0.919374</td>\n",
       "      <td>1.173146</td>\n",
       "      <td>0.549406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOMO-1</th>\n",
       "      <td>-1.542755</td>\n",
       "      <td>-1.542755</td>\n",
       "      <td>-1.542755</td>\n",
       "      <td>-1.297222</td>\n",
       "      <td>0.126352</td>\n",
       "      <td>-1.387841</td>\n",
       "      <td>0.162922</td>\n",
       "      <td>0.162922</td>\n",
       "      <td>-1.387841</td>\n",
       "      <td>-1.387841</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.542755</td>\n",
       "      <td>-1.542755</td>\n",
       "      <td>0.452033</td>\n",
       "      <td>0.453452</td>\n",
       "      <td>-0.166773</td>\n",
       "      <td>0.273399</td>\n",
       "      <td>0.972969</td>\n",
       "      <td>0.361636</td>\n",
       "      <td>0.893154</td>\n",
       "      <td>0.031334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OCI-AML2</th>\n",
       "      <td>-1.018958</td>\n",
       "      <td>-1.018958</td>\n",
       "      <td>-1.018958</td>\n",
       "      <td>-0.771221</td>\n",
       "      <td>-0.706349</td>\n",
       "      <td>-0.862653</td>\n",
       "      <td>-0.706349</td>\n",
       "      <td>-0.706349</td>\n",
       "      <td>-0.862653</td>\n",
       "      <td>-0.862653</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.018958</td>\n",
       "      <td>-1.018958</td>\n",
       "      <td>0.595761</td>\n",
       "      <td>0.821537</td>\n",
       "      <td>-1.018958</td>\n",
       "      <td>1.078511</td>\n",
       "      <td>1.430702</td>\n",
       "      <td>0.954183</td>\n",
       "      <td>1.357442</td>\n",
       "      <td>-1.018958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OCI-AML5</th>\n",
       "      <td>-1.564486</td>\n",
       "      <td>-1.564486</td>\n",
       "      <td>-1.564486</td>\n",
       "      <td>0.383411</td>\n",
       "      <td>0.327752</td>\n",
       "      <td>-1.405406</td>\n",
       "      <td>0.211047</td>\n",
       "      <td>0.211047</td>\n",
       "      <td>-1.405406</td>\n",
       "      <td>-1.405406</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.378642</td>\n",
       "      <td>0.778803</td>\n",
       "      <td>0.370341</td>\n",
       "      <td>0.551901</td>\n",
       "      <td>0.203427</td>\n",
       "      <td>1.090597</td>\n",
       "      <td>1.320302</td>\n",
       "      <td>1.000087</td>\n",
       "      <td>1.262027</td>\n",
       "      <td>0.448002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OCI-M1</th>\n",
       "      <td>-1.443504</td>\n",
       "      <td>-1.443504</td>\n",
       "      <td>0.104674</td>\n",
       "      <td>-1.203335</td>\n",
       "      <td>0.416790</td>\n",
       "      <td>-1.291974</td>\n",
       "      <td>-0.033726</td>\n",
       "      <td>-0.033726</td>\n",
       "      <td>-1.291974</td>\n",
       "      <td>-1.291974</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.443504</td>\n",
       "      <td>0.807497</td>\n",
       "      <td>0.197558</td>\n",
       "      <td>0.552184</td>\n",
       "      <td>0.121180</td>\n",
       "      <td>1.032276</td>\n",
       "      <td>1.355549</td>\n",
       "      <td>0.936137</td>\n",
       "      <td>1.307748</td>\n",
       "      <td>0.353252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OE19</th>\n",
       "      <td>0.270560</td>\n",
       "      <td>0.553191</td>\n",
       "      <td>0.823600</td>\n",
       "      <td>-0.045272</td>\n",
       "      <td>-0.044472</td>\n",
       "      <td>-1.153483</td>\n",
       "      <td>0.124525</td>\n",
       "      <td>0.124525</td>\n",
       "      <td>-1.153483</td>\n",
       "      <td>-1.153483</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.308953</td>\n",
       "      <td>-1.308953</td>\n",
       "      <td>0.390706</td>\n",
       "      <td>0.756886</td>\n",
       "      <td>0.162303</td>\n",
       "      <td>1.096679</td>\n",
       "      <td>1.302323</td>\n",
       "      <td>0.759799</td>\n",
       "      <td>1.265287</td>\n",
       "      <td>0.124380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OE33</th>\n",
       "      <td>-1.340577</td>\n",
       "      <td>0.940878</td>\n",
       "      <td>0.558403</td>\n",
       "      <td>0.148343</td>\n",
       "      <td>0.634569</td>\n",
       "      <td>-1.182143</td>\n",
       "      <td>0.320345</td>\n",
       "      <td>0.320345</td>\n",
       "      <td>-1.182143</td>\n",
       "      <td>-1.182143</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.340577</td>\n",
       "      <td>-1.340577</td>\n",
       "      <td>0.212821</td>\n",
       "      <td>0.652400</td>\n",
       "      <td>0.175780</td>\n",
       "      <td>1.247174</td>\n",
       "      <td>1.190070</td>\n",
       "      <td>0.379337</td>\n",
       "      <td>1.136030</td>\n",
       "      <td>-1.340577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P31-FUJ</th>\n",
       "      <td>-1.618652</td>\n",
       "      <td>-1.618652</td>\n",
       "      <td>-1.618652</td>\n",
       "      <td>0.390003</td>\n",
       "      <td>0.185960</td>\n",
       "      <td>-1.468356</td>\n",
       "      <td>0.003575</td>\n",
       "      <td>0.003575</td>\n",
       "      <td>0.003331</td>\n",
       "      <td>0.003331</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.618652</td>\n",
       "      <td>-1.618652</td>\n",
       "      <td>0.182697</td>\n",
       "      <td>0.578384</td>\n",
       "      <td>-0.083638</td>\n",
       "      <td>0.912580</td>\n",
       "      <td>1.192911</td>\n",
       "      <td>0.711184</td>\n",
       "      <td>1.150494</td>\n",
       "      <td>0.142746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PL-21</th>\n",
       "      <td>-1.410989</td>\n",
       "      <td>-1.410989</td>\n",
       "      <td>-1.410989</td>\n",
       "      <td>-1.155393</td>\n",
       "      <td>0.082803</td>\n",
       "      <td>-1.249726</td>\n",
       "      <td>0.124308</td>\n",
       "      <td>0.124308</td>\n",
       "      <td>-1.249726</td>\n",
       "      <td>-1.249726</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.410989</td>\n",
       "      <td>-1.410989</td>\n",
       "      <td>0.487007</td>\n",
       "      <td>0.683207</td>\n",
       "      <td>-0.122975</td>\n",
       "      <td>0.374886</td>\n",
       "      <td>1.270299</td>\n",
       "      <td>0.829148</td>\n",
       "      <td>1.196391</td>\n",
       "      <td>0.201377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIG-M5</th>\n",
       "      <td>-1.458045</td>\n",
       "      <td>-1.458045</td>\n",
       "      <td>-1.458045</td>\n",
       "      <td>0.489594</td>\n",
       "      <td>0.069848</td>\n",
       "      <td>-1.309151</td>\n",
       "      <td>-0.333655</td>\n",
       "      <td>-0.333655</td>\n",
       "      <td>0.118042</td>\n",
       "      <td>0.118042</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.458045</td>\n",
       "      <td>-1.458045</td>\n",
       "      <td>0.296312</td>\n",
       "      <td>0.440464</td>\n",
       "      <td>0.087924</td>\n",
       "      <td>1.024717</td>\n",
       "      <td>1.379408</td>\n",
       "      <td>1.009671</td>\n",
       "      <td>1.302141</td>\n",
       "      <td>0.662591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SK-HEP-1</th>\n",
       "      <td>-1.154669</td>\n",
       "      <td>0.501981</td>\n",
       "      <td>0.535036</td>\n",
       "      <td>-0.123431</td>\n",
       "      <td>-0.121410</td>\n",
       "      <td>-1.002127</td>\n",
       "      <td>-0.849585</td>\n",
       "      <td>-0.849585</td>\n",
       "      <td>-1.002127</td>\n",
       "      <td>-1.002127</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.154669</td>\n",
       "      <td>-1.154669</td>\n",
       "      <td>0.423487</td>\n",
       "      <td>0.807064</td>\n",
       "      <td>-1.154669</td>\n",
       "      <td>1.115387</td>\n",
       "      <td>1.448977</td>\n",
       "      <td>0.927273</td>\n",
       "      <td>1.410489</td>\n",
       "      <td>0.585575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SKM-1</th>\n",
       "      <td>-1.738019</td>\n",
       "      <td>-1.738019</td>\n",
       "      <td>-1.738019</td>\n",
       "      <td>0.153588</td>\n",
       "      <td>-1.411161</td>\n",
       "      <td>-1.574590</td>\n",
       "      <td>-1.411161</td>\n",
       "      <td>-1.411161</td>\n",
       "      <td>-1.574590</td>\n",
       "      <td>-1.574590</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.166520</td>\n",
       "      <td>0.960051</td>\n",
       "      <td>0.249349</td>\n",
       "      <td>0.469887</td>\n",
       "      <td>-0.307177</td>\n",
       "      <td>0.875424</td>\n",
       "      <td>1.196133</td>\n",
       "      <td>0.725343</td>\n",
       "      <td>1.151595</td>\n",
       "      <td>0.117719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNU-387</th>\n",
       "      <td>0.056274</td>\n",
       "      <td>0.062127</td>\n",
       "      <td>-1.215612</td>\n",
       "      <td>-0.967579</td>\n",
       "      <td>-0.902630</td>\n",
       "      <td>-1.059121</td>\n",
       "      <td>0.162760</td>\n",
       "      <td>0.162760</td>\n",
       "      <td>0.161750</td>\n",
       "      <td>0.161750</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.215612</td>\n",
       "      <td>-1.215612</td>\n",
       "      <td>-0.107164</td>\n",
       "      <td>0.538494</td>\n",
       "      <td>-1.215612</td>\n",
       "      <td>0.952772</td>\n",
       "      <td>1.258943</td>\n",
       "      <td>0.692819</td>\n",
       "      <td>1.225107</td>\n",
       "      <td>0.232410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNU-398</th>\n",
       "      <td>-0.988966</td>\n",
       "      <td>1.318189</td>\n",
       "      <td>-0.988966</td>\n",
       "      <td>0.981400</td>\n",
       "      <td>-0.673947</td>\n",
       "      <td>-0.831457</td>\n",
       "      <td>0.391071</td>\n",
       "      <td>0.391071</td>\n",
       "      <td>-0.831457</td>\n",
       "      <td>-0.831457</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.988966</td>\n",
       "      <td>-0.988966</td>\n",
       "      <td>1.284576</td>\n",
       "      <td>1.476408</td>\n",
       "      <td>-0.988966</td>\n",
       "      <td>1.569048</td>\n",
       "      <td>1.312939</td>\n",
       "      <td>-0.831457</td>\n",
       "      <td>1.228888</td>\n",
       "      <td>-0.988966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNU-423</th>\n",
       "      <td>-1.049877</td>\n",
       "      <td>0.786786</td>\n",
       "      <td>-1.049877</td>\n",
       "      <td>1.121893</td>\n",
       "      <td>0.167315</td>\n",
       "      <td>-0.891829</td>\n",
       "      <td>0.185983</td>\n",
       "      <td>0.185983</td>\n",
       "      <td>-0.891829</td>\n",
       "      <td>-0.891829</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.049877</td>\n",
       "      <td>-1.049877</td>\n",
       "      <td>0.654917</td>\n",
       "      <td>0.985332</td>\n",
       "      <td>-1.049877</td>\n",
       "      <td>1.270787</td>\n",
       "      <td>1.339983</td>\n",
       "      <td>-0.891829</td>\n",
       "      <td>1.289666</td>\n",
       "      <td>-1.049877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNU-449</th>\n",
       "      <td>-1.303739</td>\n",
       "      <td>0.663216</td>\n",
       "      <td>-0.037737</td>\n",
       "      <td>0.434663</td>\n",
       "      <td>0.826194</td>\n",
       "      <td>-1.158293</td>\n",
       "      <td>0.197001</td>\n",
       "      <td>0.197001</td>\n",
       "      <td>-1.158293</td>\n",
       "      <td>-1.158293</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.303739</td>\n",
       "      <td>-1.303739</td>\n",
       "      <td>0.115506</td>\n",
       "      <td>0.639408</td>\n",
       "      <td>-1.303739</td>\n",
       "      <td>1.073104</td>\n",
       "      <td>1.283087</td>\n",
       "      <td>0.784328</td>\n",
       "      <td>1.249925</td>\n",
       "      <td>0.317130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNU-475</th>\n",
       "      <td>0.318143</td>\n",
       "      <td>0.208604</td>\n",
       "      <td>-1.244337</td>\n",
       "      <td>0.164870</td>\n",
       "      <td>0.554723</td>\n",
       "      <td>-1.100630</td>\n",
       "      <td>-0.956923</td>\n",
       "      <td>-0.956923</td>\n",
       "      <td>-1.100630</td>\n",
       "      <td>-1.100630</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.244337</td>\n",
       "      <td>-1.244337</td>\n",
       "      <td>0.269708</td>\n",
       "      <td>0.572032</td>\n",
       "      <td>-1.244337</td>\n",
       "      <td>0.976678</td>\n",
       "      <td>1.175560</td>\n",
       "      <td>0.667882</td>\n",
       "      <td>1.146065</td>\n",
       "      <td>-1.244337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THP-1</th>\n",
       "      <td>-1.633713</td>\n",
       "      <td>-1.633713</td>\n",
       "      <td>0.091243</td>\n",
       "      <td>0.328972</td>\n",
       "      <td>0.041054</td>\n",
       "      <td>-1.492195</td>\n",
       "      <td>-0.018010</td>\n",
       "      <td>-0.018010</td>\n",
       "      <td>-0.018160</td>\n",
       "      <td>-0.018160</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.633713</td>\n",
       "      <td>0.841201</td>\n",
       "      <td>0.573624</td>\n",
       "      <td>0.504417</td>\n",
       "      <td>-1.633713</td>\n",
       "      <td>0.891386</td>\n",
       "      <td>1.206627</td>\n",
       "      <td>1.010553</td>\n",
       "      <td>1.122727</td>\n",
       "      <td>0.750943</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38 rows × 1064 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "col.name    EGFR(T693);  EGFR(S695);  EGFR(S1166);  EGFR(S991);  EGFR(T993);  \\\n",
       "COLO-680N      0.870410     0.859884      0.487725     0.522972     0.688719   \n",
       "GDM-1         -1.429010    -1.429010     -1.429010     0.430902     0.578810   \n",
       "HEL           -1.613351    -1.613351     -1.613351     0.336451     0.457029   \n",
       "HL-60          0.426137    -1.428863     -1.428863    -1.206791     0.335992   \n",
       "JHH-2          0.713488     0.820889      0.532149     0.686185     1.006834   \n",
       "JHH-4          0.770449     0.857518      0.400208     0.617482     0.778469   \n",
       "KASUMI-1      -1.316937    -1.316937     -1.316937    -1.091034    -1.031879   \n",
       "KG-1          -1.570566    -1.570566     -0.100433     0.408062    -1.263924   \n",
       "KMOE-2        -0.204030    -1.659307      0.168107     0.304357     0.350523   \n",
       "KYSE-140      -1.231059    -1.231059      0.144506     0.192449    -0.913721   \n",
       "KYSE-150      -1.259273     0.570344     -1.259273     0.555773     0.675030   \n",
       "KYSE-410      -1.307672     0.314822      0.757654     0.457031    -0.985960   \n",
       "KYSE-450       0.857134     0.980034      0.208004     0.416529     0.349901   \n",
       "KYSE-510       0.924416    -0.846311     -0.846311    -0.608318    -0.545998   \n",
       "KYSE-520       0.364092     0.691327      0.587081     0.421782    -0.863160   \n",
       "KYSE-70       -0.051558     0.269311     -1.335412     0.168138     0.513250   \n",
       "ME-1          -1.543695    -1.543695     -0.007512    -1.319835     0.401082   \n",
       "ML-2          -1.598369    -1.598369     -1.598369    -1.365840    -0.040885   \n",
       "MOLM-13        0.256812     0.061433     -1.511135     0.282191     0.381162   \n",
       "MONO-MAC-6    -1.652866    -1.652866     -1.652866    -1.433589     0.079342   \n",
       "MV-4-11       -0.429467    -1.530044     -1.530044     0.276384     0.322332   \n",
       "NOMO-1        -1.542755    -1.542755     -1.542755    -1.297222     0.126352   \n",
       "OCI-AML2      -1.018958    -1.018958     -1.018958    -0.771221    -0.706349   \n",
       "OCI-AML5      -1.564486    -1.564486     -1.564486     0.383411     0.327752   \n",
       "OCI-M1        -1.443504    -1.443504      0.104674    -1.203335     0.416790   \n",
       "OE19           0.270560     0.553191      0.823600    -0.045272    -0.044472   \n",
       "OE33          -1.340577     0.940878      0.558403     0.148343     0.634569   \n",
       "P31-FUJ       -1.618652    -1.618652     -1.618652     0.390003     0.185960   \n",
       "PL-21         -1.410989    -1.410989     -1.410989    -1.155393     0.082803   \n",
       "SIG-M5        -1.458045    -1.458045     -1.458045     0.489594     0.069848   \n",
       "SK-HEP-1      -1.154669     0.501981      0.535036    -0.123431    -0.121410   \n",
       "SKM-1         -1.738019    -1.738019     -1.738019     0.153588    -1.411161   \n",
       "SNU-387        0.056274     0.062127     -1.215612    -0.967579    -0.902630   \n",
       "SNU-398       -0.988966     1.318189     -0.988966     0.981400    -0.673947   \n",
       "SNU-423       -1.049877     0.786786     -1.049877     1.121893     0.167315   \n",
       "SNU-449       -1.303739     0.663216     -0.037737     0.434663     0.826194   \n",
       "SNU-475        0.318143     0.208604     -1.244337     0.164870     0.554723   \n",
       "THP-1         -1.633713    -1.633713      0.091243     0.328972     0.041054   \n",
       "\n",
       "col.name    EGFR(S995);  EGFR(T1032);EGFR(T1046);  EGFR(T1046);  \\\n",
       "COLO-680N     -1.130207                  0.948421      0.948421   \n",
       "GDM-1         -1.271041                 -0.048920     -0.048920   \n",
       "HEL           -1.463476                 -1.313602     -1.313602   \n",
       "HL-60         -1.288751                  0.373059      0.373059   \n",
       "JHH-2         -1.076954                  0.530054      0.530054   \n",
       "JHH-4          0.272771                  0.784972      0.784972   \n",
       "KASUMI-1      -1.174408                  0.209470      0.209470   \n",
       "KG-1          -1.417245                 -1.263924     -1.263924   \n",
       "KMOE-2        -1.502197                 -0.092908     -0.092908   \n",
       "KYSE-140      -1.072390                  0.553152      0.553152   \n",
       "KYSE-150      -1.104834                  0.499938      0.499938   \n",
       "KYSE-410      -1.146816                 -0.985960     -0.985960   \n",
       "KYSE-450      -1.081212                  0.300771      0.300771   \n",
       "KYSE-510      -0.696154                 -0.545998     -0.545998   \n",
       "KYSE-520      -1.025986                  0.284347      0.284347   \n",
       "KYSE-70       -1.175207                 -0.037395     -0.037395   \n",
       "ME-1          -1.402455                 -1.261215     -1.261215   \n",
       "ML-2          -1.451659                  0.353061      0.353061   \n",
       "MOLM-13       -1.360741                 -1.210347     -1.210347   \n",
       "MONO-MAC-6    -1.514517                 -1.376169     -1.376169   \n",
       "MV-4-11       -1.371465                  0.205197      0.205197   \n",
       "NOMO-1        -1.387841                  0.162922      0.162922   \n",
       "OCI-AML2      -0.862653                 -0.706349     -0.706349   \n",
       "OCI-AML5      -1.405406                  0.211047      0.211047   \n",
       "OCI-M1        -1.291974                 -0.033726     -0.033726   \n",
       "OE19          -1.153483                  0.124525      0.124525   \n",
       "OE33          -1.182143                  0.320345      0.320345   \n",
       "P31-FUJ       -1.468356                  0.003575      0.003575   \n",
       "PL-21         -1.249726                  0.124308      0.124308   \n",
       "SIG-M5        -1.309151                 -0.333655     -0.333655   \n",
       "SK-HEP-1      -1.002127                 -0.849585     -0.849585   \n",
       "SKM-1         -1.574590                 -1.411161     -1.411161   \n",
       "SNU-387       -1.059121                  0.162760      0.162760   \n",
       "SNU-398       -0.831457                  0.391071      0.391071   \n",
       "SNU-423       -0.891829                  0.185983      0.185983   \n",
       "SNU-449       -1.158293                  0.197001      0.197001   \n",
       "SNU-475       -1.100630                 -0.956923     -0.956923   \n",
       "THP-1         -1.492195                 -0.018010     -0.018010   \n",
       "\n",
       "col.name    EGFR(S1042);EGFR(S1045);  EGFR(S1045);  ...  EML3(S157);  \\\n",
       "COLO-680N                  -1.130207     -1.130207  ...    -1.287052   \n",
       "GDM-1                      -1.271041     -1.271041  ...    -1.429010   \n",
       "HEL                        -1.463476     -1.463476  ...    -1.613351   \n",
       "HL-60                       0.403272      0.403272  ...    -1.428863   \n",
       "JHH-2                      -1.076954     -1.076954  ...    -1.224318   \n",
       "JHH-4                       0.338587      0.338587  ...    -1.273965   \n",
       "KASUMI-1                    0.408179      0.408179  ...     0.184882   \n",
       "KG-1                       -1.417245     -1.417245  ...    -1.570566   \n",
       "KMOE-2                     -1.502197     -1.502197  ...    -0.144783   \n",
       "KYSE-140                   -1.072390     -1.072390  ...    -1.231059   \n",
       "KYSE-150                   -1.104834     -1.104834  ...    -1.259273   \n",
       "KYSE-410                   -1.146816     -1.146816  ...    -1.307672   \n",
       "KYSE-450                   -1.081212     -1.081212  ...    -1.244082   \n",
       "KYSE-510                   -0.696154     -0.696154  ...    -0.846311   \n",
       "KYSE-520                   -1.025986     -1.025986  ...    -1.188813   \n",
       "KYSE-70                    -1.175207     -1.175207  ...    -1.335412   \n",
       "ME-1                       -1.402455     -1.402455  ...    -1.543695   \n",
       "ML-2                        0.030200      0.030200  ...    -1.598369   \n",
       "MOLM-13                    -1.360741     -1.360741  ...    -0.096317   \n",
       "MONO-MAC-6                 -1.514517     -1.514517  ...    -1.652866   \n",
       "MV-4-11                    -1.371465     -1.371465  ...    -1.530044   \n",
       "NOMO-1                     -1.387841     -1.387841  ...    -1.542755   \n",
       "OCI-AML2                   -0.862653     -0.862653  ...    -1.018958   \n",
       "OCI-AML5                   -1.405406     -1.405406  ...    -0.378642   \n",
       "OCI-M1                     -1.291974     -1.291974  ...    -1.443504   \n",
       "OE19                       -1.153483     -1.153483  ...    -1.308953   \n",
       "OE33                       -1.182143     -1.182143  ...    -1.340577   \n",
       "P31-FUJ                     0.003331      0.003331  ...    -1.618652   \n",
       "PL-21                      -1.249726     -1.249726  ...    -1.410989   \n",
       "SIG-M5                      0.118042      0.118042  ...    -1.458045   \n",
       "SK-HEP-1                   -1.002127     -1.002127  ...    -1.154669   \n",
       "SKM-1                      -1.574590     -1.574590  ...    -0.166520   \n",
       "SNU-387                     0.161750      0.161750  ...    -1.215612   \n",
       "SNU-398                    -0.831457     -0.831457  ...    -0.988966   \n",
       "SNU-423                    -0.891829     -0.891829  ...    -1.049877   \n",
       "SNU-449                    -1.158293     -1.158293  ...    -1.303739   \n",
       "SNU-475                    -1.100630     -1.100630  ...    -1.244337   \n",
       "THP-1                      -0.018160     -0.018160  ...    -1.633713   \n",
       "\n",
       "col.name    EML3(S156);  EML3(S889);  EML3(S883);  EZH2(T388);  EZH2(T487);  \\\n",
       "COLO-680N     -1.287052    -1.287052     0.627996    -1.287052     1.059854   \n",
       "GDM-1          0.754475     0.288527     0.462040     0.185137     1.045595   \n",
       "HEL           -1.613351     0.349405     0.773424    -1.613351     0.970619   \n",
       "HL-60         -1.428863     0.357729     0.508261    -1.428863     0.727035   \n",
       "JHH-2         -1.224318     0.607585     0.410201    -1.224318     0.906618   \n",
       "JHH-4         -1.273965     0.362711     0.591988    -1.273965     1.174851   \n",
       "KASUMI-1      -1.316937     0.650288     0.609231     0.299954     0.797801   \n",
       "KG-1           0.770972     0.431089     0.649533    -1.570566     0.890752   \n",
       "KMOE-2         0.784860     0.222212     0.502510     0.156714     1.072007   \n",
       "KYSE-140      -1.231059    -1.231059     0.503357     0.507232     1.240663   \n",
       "KYSE-150      -1.259273     0.422692     0.653158     0.260699     1.073071   \n",
       "KYSE-410      -1.307672     0.215488     0.671119     0.207402     1.156456   \n",
       "KYSE-450      -1.244082     0.593143     0.915016    -1.244082     1.160151   \n",
       "KYSE-510      -0.846311    -0.846311    -0.846311    -0.846311     1.303020   \n",
       "KYSE-520      -1.188813     0.239823     0.769691    -1.188813     1.326737   \n",
       "KYSE-70       -1.335412    -1.335412     0.441229    -0.389097     1.025841   \n",
       "ME-1           0.142679     0.493540     0.499472     0.065131     1.084595   \n",
       "ML-2           0.385900     0.297178     0.448227    -0.199134     0.925116   \n",
       "MOLM-13        0.889257     0.453363     0.565244     0.089339     0.344541   \n",
       "MONO-MAC-6    -0.845395     0.517394     0.634201    -0.249320     0.439550   \n",
       "MV-4-11       -1.530044     0.577106     0.632136     0.255899     1.101878   \n",
       "NOMO-1        -1.542755     0.452033     0.453452    -0.166773     0.273399   \n",
       "OCI-AML2      -1.018958     0.595761     0.821537    -1.018958     1.078511   \n",
       "OCI-AML5       0.778803     0.370341     0.551901     0.203427     1.090597   \n",
       "OCI-M1         0.807497     0.197558     0.552184     0.121180     1.032276   \n",
       "OE19          -1.308953     0.390706     0.756886     0.162303     1.096679   \n",
       "OE33          -1.340577     0.212821     0.652400     0.175780     1.247174   \n",
       "P31-FUJ       -1.618652     0.182697     0.578384    -0.083638     0.912580   \n",
       "PL-21         -1.410989     0.487007     0.683207    -0.122975     0.374886   \n",
       "SIG-M5        -1.458045     0.296312     0.440464     0.087924     1.024717   \n",
       "SK-HEP-1      -1.154669     0.423487     0.807064    -1.154669     1.115387   \n",
       "SKM-1          0.960051     0.249349     0.469887    -0.307177     0.875424   \n",
       "SNU-387       -1.215612    -0.107164     0.538494    -1.215612     0.952772   \n",
       "SNU-398       -0.988966     1.284576     1.476408    -0.988966     1.569048   \n",
       "SNU-423       -1.049877     0.654917     0.985332    -1.049877     1.270787   \n",
       "SNU-449       -1.303739     0.115506     0.639408    -1.303739     1.073104   \n",
       "SNU-475       -1.244337     0.269708     0.572032    -1.244337     0.976678   \n",
       "THP-1          0.841201     0.573624     0.504417    -1.633713     0.891386   \n",
       "\n",
       "col.name    EZH2(S362);  EZH2(T369);  EZH2(S362);EZH2(S363);  EZH2(S366);  \n",
       "COLO-680N      1.246416     0.640216                1.215685    -1.287052  \n",
       "GDM-1          1.414728     0.925482                1.374856    -1.429010  \n",
       "HEL            1.327014     0.845071                1.271451     0.492025  \n",
       "HL-60          1.430922     0.995132                1.396240     0.468569  \n",
       "JHH-2          1.055518     0.600364                1.031041    -1.224318  \n",
       "JHH-4          1.317839     0.716005                1.273343     0.322767  \n",
       "KASUMI-1       1.375823     1.048650                1.319048    -1.316937  \n",
       "KG-1           1.019903     0.291344                0.984688     0.126431  \n",
       "KMOE-2         1.310393     0.838632                1.248473     0.521745  \n",
       "KYSE-140       1.300372     0.752033                1.251735    -1.231059  \n",
       "KYSE-150       1.249470     0.706562                1.217383    -1.259273  \n",
       "KYSE-410       1.230030     0.547308                1.198218    -1.307672  \n",
       "KYSE-450       1.153924     0.406590                1.100142    -1.244082  \n",
       "KYSE-510       1.618310     1.151798                1.591581    -0.846311  \n",
       "KYSE-520       1.276652     0.532564                1.193026     0.341449  \n",
       "KYSE-70        1.172570     0.575567                1.134710    -1.335412  \n",
       "ME-1           1.371463     0.951912                1.330122     0.493238  \n",
       "ML-2           1.231671     1.023237                1.154168    -1.598369  \n",
       "MOLM-13        1.249249     0.769004                1.161006     0.461591  \n",
       "MONO-MAC-6     1.075657     0.847440                0.976697     0.486960  \n",
       "MV-4-11        1.252620     0.919374                1.173146     0.549406  \n",
       "NOMO-1         0.972969     0.361636                0.893154     0.031334  \n",
       "OCI-AML2       1.430702     0.954183                1.357442    -1.018958  \n",
       "OCI-AML5       1.320302     1.000087                1.262027     0.448002  \n",
       "OCI-M1         1.355549     0.936137                1.307748     0.353252  \n",
       "OE19           1.302323     0.759799                1.265287     0.124380  \n",
       "OE33           1.190070     0.379337                1.136030    -1.340577  \n",
       "P31-FUJ        1.192911     0.711184                1.150494     0.142746  \n",
       "PL-21          1.270299     0.829148                1.196391     0.201377  \n",
       "SIG-M5         1.379408     1.009671                1.302141     0.662591  \n",
       "SK-HEP-1       1.448977     0.927273                1.410489     0.585575  \n",
       "SKM-1          1.196133     0.725343                1.151595     0.117719  \n",
       "SNU-387        1.258943     0.692819                1.225107     0.232410  \n",
       "SNU-398        1.312939    -0.831457                1.228888    -0.988966  \n",
       "SNU-423        1.339983    -0.891829                1.289666    -1.049877  \n",
       "SNU-449        1.283087     0.784328                1.249925     0.317130  \n",
       "SNU-475        1.175560     0.667882                1.146065    -1.244337  \n",
       "THP-1          1.206627     1.010553                1.122727     0.750943  \n",
       "\n",
       "[38 rows x 1064 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phos_genes = np.array([c.split('(')[0] for c in phospho_ls.columns])\n",
    "unique_phos = np.unique(phos_genes)\n",
    "overlapping_landmarks = [g for g in landmark_genes.index if g in unique_phos]\n",
    "\n",
    "#read in targets of proteins data\n",
    "dd_path = '/data/home/wpw035/down_data_small'\n",
    "enz_sub = pd.read_csv(f'{dd_path}/enzsub_relations_phosSite_signor.csv', index_col=0)\n",
    "enz_sub.index = enz_sub['enzyme_genesymbol']\n",
    "\n",
    "#landmark genes with targets\n",
    "lands_with_target = [g for g in landmark_genes.index if g in enz_sub.index]\n",
    "land_targets = set(enz_sub.loc[lands_with_target]['substrate_genesymbol'])\n",
    "#landmark genes with targets that are in phos data\n",
    "overlapping_land_targets = [t for t in land_targets if t in unique_phos]\n",
    "\n",
    "targets_also_landmarks = set(overlapping_land_targets).intersection(\n",
    "    landmark_genes.index) \n",
    "print(len(targets_also_landmarks))\n",
    "\n",
    "overlap_inds = []\n",
    "for gene in targets_also_landmarks:\n",
    "    overlap_inds.extend(np.where(phos_genes == gene)[0])\n",
    "print(len(overlap_inds))\n",
    "\n",
    "kep_cols = phospho_ls.columns[overlap_inds]\n",
    "phospho_ls[kep_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Drug cell line input data creation.\n",
    "\n",
    "Creates inputs for each drug cell line pair using the create_all_drugs fuction,\n",
    "have a look in the utils module to see the create_all_drugs function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((11583, 908), (11583, 345), 11583)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create input data for each drug\n",
    "x_all, x_drug, y_series = utils.create_all_drugs(\n",
    "    rna, one_hot_drugs, ic50_df1, _all_cls)\n",
    "\n",
    "x_all = x_all.astype(np.float32)\n",
    "x_drug = x_drug.astype(np.float16)\n",
    "\n",
    "#fmt index to include drug cell line paris\n",
    "cls_drugs_index = x_all.index + '::' + x_drug.index\n",
    "x_all.index = cls_drugs_index\n",
    "x_drug.index = cls_drugs_index\n",
    "y_series.index = cls_drugs_index\n",
    "\n",
    "x_all.shape, x_drug.shape, len(y_series)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have three objects for use in our models. Two input dataframes, x_all and x_drug. x_all gives the rna-seq data for each drug cell line pair, and x_drug gives the one-hot representation of the drugs. y_series is a pandas series object that gives the target IC50 values for each drug cell line pair. \n",
    "\n",
    "Notice that the index of the dataframes match up with the target values. Such that x_all.iloc[i], x_drug.iloc[i] and y_series[i] refer to the same drug cell line pair. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GENE_SYMBOLS</th>\n",
       "      <th>INSIG1</th>\n",
       "      <th>FOXO3</th>\n",
       "      <th>CDH3</th>\n",
       "      <th>ORC1</th>\n",
       "      <th>ITGAE</th>\n",
       "      <th>CDK7</th>\n",
       "      <th>FOS</th>\n",
       "      <th>CDK6</th>\n",
       "      <th>CDK4</th>\n",
       "      <th>CDK2</th>\n",
       "      <th>...</th>\n",
       "      <th>FHL2</th>\n",
       "      <th>IL13RA1</th>\n",
       "      <th>FKBP4</th>\n",
       "      <th>ABHD4</th>\n",
       "      <th>HIST1H2BK</th>\n",
       "      <th>NUCB2</th>\n",
       "      <th>NUP88</th>\n",
       "      <th>NVL</th>\n",
       "      <th>ILK</th>\n",
       "      <th>INPP1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>COLO-680N::(5Z)-7-Oxozeaenol</th>\n",
       "      <td>8.154784</td>\n",
       "      <td>9.236981</td>\n",
       "      <td>9.799767</td>\n",
       "      <td>5.103765</td>\n",
       "      <td>10.757260</td>\n",
       "      <td>9.690476</td>\n",
       "      <td>9.034679</td>\n",
       "      <td>6.728485</td>\n",
       "      <td>8.812278</td>\n",
       "      <td>6.160667</td>\n",
       "      <td>...</td>\n",
       "      <td>8.719871</td>\n",
       "      <td>6.707273</td>\n",
       "      <td>7.683773</td>\n",
       "      <td>3.787795</td>\n",
       "      <td>9.465148</td>\n",
       "      <td>7.516577</td>\n",
       "      <td>7.447703</td>\n",
       "      <td>6.983274</td>\n",
       "      <td>6.795911</td>\n",
       "      <td>9.728236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDM-1::(5Z)-7-Oxozeaenol</th>\n",
       "      <td>8.055311</td>\n",
       "      <td>8.088824</td>\n",
       "      <td>3.379974</td>\n",
       "      <td>6.465721</td>\n",
       "      <td>9.480014</td>\n",
       "      <td>8.702067</td>\n",
       "      <td>3.157377</td>\n",
       "      <td>9.285379</td>\n",
       "      <td>8.979623</td>\n",
       "      <td>6.235116</td>\n",
       "      <td>...</td>\n",
       "      <td>6.329245</td>\n",
       "      <td>4.010940</td>\n",
       "      <td>9.105475</td>\n",
       "      <td>3.394924</td>\n",
       "      <td>8.919325</td>\n",
       "      <td>8.434682</td>\n",
       "      <td>7.330557</td>\n",
       "      <td>6.003414</td>\n",
       "      <td>6.626827</td>\n",
       "      <td>8.023713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HEL::(5Z)-7-Oxozeaenol</th>\n",
       "      <td>7.678645</td>\n",
       "      <td>8.126147</td>\n",
       "      <td>3.383949</td>\n",
       "      <td>7.070015</td>\n",
       "      <td>8.819721</td>\n",
       "      <td>8.586131</td>\n",
       "      <td>6.420190</td>\n",
       "      <td>8.927892</td>\n",
       "      <td>9.720554</td>\n",
       "      <td>6.398864</td>\n",
       "      <td>...</td>\n",
       "      <td>9.807980</td>\n",
       "      <td>3.422834</td>\n",
       "      <td>7.842642</td>\n",
       "      <td>3.879684</td>\n",
       "      <td>11.686403</td>\n",
       "      <td>7.181875</td>\n",
       "      <td>6.733916</td>\n",
       "      <td>6.101298</td>\n",
       "      <td>6.616129</td>\n",
       "      <td>9.701949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HL-60::(5Z)-7-Oxozeaenol</th>\n",
       "      <td>9.919567</td>\n",
       "      <td>7.989809</td>\n",
       "      <td>3.171409</td>\n",
       "      <td>7.077472</td>\n",
       "      <td>8.133847</td>\n",
       "      <td>8.347510</td>\n",
       "      <td>6.516647</td>\n",
       "      <td>8.619617</td>\n",
       "      <td>9.546260</td>\n",
       "      <td>6.068342</td>\n",
       "      <td>...</td>\n",
       "      <td>3.166125</td>\n",
       "      <td>4.936508</td>\n",
       "      <td>7.779411</td>\n",
       "      <td>3.562882</td>\n",
       "      <td>11.235920</td>\n",
       "      <td>10.315398</td>\n",
       "      <td>6.815571</td>\n",
       "      <td>6.391719</td>\n",
       "      <td>6.212755</td>\n",
       "      <td>5.730575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JHH-2::(5Z)-7-Oxozeaenol</th>\n",
       "      <td>9.251225</td>\n",
       "      <td>8.227397</td>\n",
       "      <td>3.260621</td>\n",
       "      <td>3.861217</td>\n",
       "      <td>9.561527</td>\n",
       "      <td>8.988581</td>\n",
       "      <td>5.048408</td>\n",
       "      <td>8.471914</td>\n",
       "      <td>8.982768</td>\n",
       "      <td>5.421257</td>\n",
       "      <td>...</td>\n",
       "      <td>9.257870</td>\n",
       "      <td>8.454283</td>\n",
       "      <td>6.429326</td>\n",
       "      <td>5.200694</td>\n",
       "      <td>9.746995</td>\n",
       "      <td>8.678383</td>\n",
       "      <td>6.247309</td>\n",
       "      <td>6.080807</td>\n",
       "      <td>7.977128</td>\n",
       "      <td>8.423243</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 908 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GENE_SYMBOLS                    INSIG1     FOXO3      CDH3      ORC1  \\\n",
       "COLO-680N::(5Z)-7-Oxozeaenol  8.154784  9.236981  9.799767  5.103765   \n",
       "GDM-1::(5Z)-7-Oxozeaenol      8.055311  8.088824  3.379974  6.465721   \n",
       "HEL::(5Z)-7-Oxozeaenol        7.678645  8.126147  3.383949  7.070015   \n",
       "HL-60::(5Z)-7-Oxozeaenol      9.919567  7.989809  3.171409  7.077472   \n",
       "JHH-2::(5Z)-7-Oxozeaenol      9.251225  8.227397  3.260621  3.861217   \n",
       "\n",
       "GENE_SYMBOLS                      ITGAE      CDK7       FOS      CDK6  \\\n",
       "COLO-680N::(5Z)-7-Oxozeaenol  10.757260  9.690476  9.034679  6.728485   \n",
       "GDM-1::(5Z)-7-Oxozeaenol       9.480014  8.702067  3.157377  9.285379   \n",
       "HEL::(5Z)-7-Oxozeaenol         8.819721  8.586131  6.420190  8.927892   \n",
       "HL-60::(5Z)-7-Oxozeaenol       8.133847  8.347510  6.516647  8.619617   \n",
       "JHH-2::(5Z)-7-Oxozeaenol       9.561527  8.988581  5.048408  8.471914   \n",
       "\n",
       "GENE_SYMBOLS                      CDK4      CDK2  ...      FHL2   IL13RA1  \\\n",
       "COLO-680N::(5Z)-7-Oxozeaenol  8.812278  6.160667  ...  8.719871  6.707273   \n",
       "GDM-1::(5Z)-7-Oxozeaenol      8.979623  6.235116  ...  6.329245  4.010940   \n",
       "HEL::(5Z)-7-Oxozeaenol        9.720554  6.398864  ...  9.807980  3.422834   \n",
       "HL-60::(5Z)-7-Oxozeaenol      9.546260  6.068342  ...  3.166125  4.936508   \n",
       "JHH-2::(5Z)-7-Oxozeaenol      8.982768  5.421257  ...  9.257870  8.454283   \n",
       "\n",
       "GENE_SYMBOLS                     FKBP4     ABHD4  HIST1H2BK      NUCB2  \\\n",
       "COLO-680N::(5Z)-7-Oxozeaenol  7.683773  3.787795   9.465148   7.516577   \n",
       "GDM-1::(5Z)-7-Oxozeaenol      9.105475  3.394924   8.919325   8.434682   \n",
       "HEL::(5Z)-7-Oxozeaenol        7.842642  3.879684  11.686403   7.181875   \n",
       "HL-60::(5Z)-7-Oxozeaenol      7.779411  3.562882  11.235920  10.315398   \n",
       "JHH-2::(5Z)-7-Oxozeaenol      6.429326  5.200694   9.746995   8.678383   \n",
       "\n",
       "GENE_SYMBOLS                     NUP88       NVL       ILK     INPP1  \n",
       "COLO-680N::(5Z)-7-Oxozeaenol  7.447703  6.983274  6.795911  9.728236  \n",
       "GDM-1::(5Z)-7-Oxozeaenol      7.330557  6.003414  6.626827  8.023713  \n",
       "HEL::(5Z)-7-Oxozeaenol        6.733916  6.101298  6.616129  9.701949  \n",
       "HL-60::(5Z)-7-Oxozeaenol      6.815571  6.391719  6.212755  5.730575  \n",
       "JHH-2::(5Z)-7-Oxozeaenol      6.247309  6.080807  7.977128  8.423243  \n",
       "\n",
       "[5 rows x 908 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>335</th>\n",
       "      <th>336</th>\n",
       "      <th>337</th>\n",
       "      <th>338</th>\n",
       "      <th>339</th>\n",
       "      <th>340</th>\n",
       "      <th>341</th>\n",
       "      <th>342</th>\n",
       "      <th>343</th>\n",
       "      <th>344</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>COLO-680N::(5Z)-7-Oxozeaenol</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDM-1::(5Z)-7-Oxozeaenol</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HEL::(5Z)-7-Oxozeaenol</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HL-60::(5Z)-7-Oxozeaenol</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JHH-2::(5Z)-7-Oxozeaenol</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 345 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              0    1    2    3    4    5    6    7    8    \\\n",
       "COLO-680N::(5Z)-7-Oxozeaenol  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "GDM-1::(5Z)-7-Oxozeaenol      1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "HEL::(5Z)-7-Oxozeaenol        1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "HL-60::(5Z)-7-Oxozeaenol      1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "JHH-2::(5Z)-7-Oxozeaenol      1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "\n",
       "                              9    ...  335  336  337  338  339  340  341  \\\n",
       "COLO-680N::(5Z)-7-Oxozeaenol  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "GDM-1::(5Z)-7-Oxozeaenol      0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "HEL::(5Z)-7-Oxozeaenol        0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "HL-60::(5Z)-7-Oxozeaenol      0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "JHH-2::(5Z)-7-Oxozeaenol      0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "\n",
       "                              342  343  344  \n",
       "COLO-680N::(5Z)-7-Oxozeaenol  0.0  0.0  0.0  \n",
       "GDM-1::(5Z)-7-Oxozeaenol      0.0  0.0  0.0  \n",
       "HEL::(5Z)-7-Oxozeaenol        0.0  0.0  0.0  \n",
       "HL-60::(5Z)-7-Oxozeaenol      0.0  0.0  0.0  \n",
       "JHH-2::(5Z)-7-Oxozeaenol      0.0  0.0  0.0  \n",
       "\n",
       "[5 rows x 345 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_drug.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "COLO-680N::(5Z)-7-Oxozeaenol    1.729743\n",
       "GDM-1::(5Z)-7-Oxozeaenol       -0.523250\n",
       "HEL::(5Z)-7-Oxozeaenol          0.948509\n",
       "HL-60::(5Z)-7-Oxozeaenol        1.426720\n",
       "JHH-2::(5Z)-7-Oxozeaenol        1.250863\n",
       "dtype: float64"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_series.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test train split\n",
    "\n",
    "using split method from my splitting module in my DRP_nb library, which allows for different splitting types to be used. Here use the cancer blind (c-blind) splitting type. Cancer blind splitting means the cell lines in the testing set are not in the training and vice versa."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fraction of cls in sets, relative to all clsbefore mising values are removed\n",
      "train fraction 0.5789473684210527, test fraction 0.42105263157894735\n",
      "------\n",
      "Fraction of cls in sets, relative to all cl drug pairs, after mising values are removed\n",
      "train fraction 0.5201372997711671, test fraction 0.36338672768878716\n"
     ]
    }
   ],
   "source": [
    "#test train split indexes\n",
    "rand_seed = 42\n",
    "_train_size = 0.6\n",
    "pairs_with_truth_vals =  y_series.index\n",
    "train_pairs, test_pairs = splitting.split(\n",
    "    rand_seed, _all_cls, _all_drugs, pairs_with_truth_vals,\n",
    "    train_size=_train_size, split_type='cblind')\n",
    "\n",
    "#test train selection\n",
    "#omics dfs\n",
    "xo_train_rna, xo_test_rna = x_all.loc[train_pairs], x_all.loc[test_pairs] \n",
    "xd_train, xd_test = x_drug.loc[train_pairs], x_drug.loc[test_pairs] #drug dfs\n",
    "y_train, y_test = y_series[train_pairs], y_series[test_pairs] #target series"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((6819, 908), (6819, 345), (6819,))"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xo_train_rna.shape, xd_train.shape, y_train.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((4764, 908), (4764, 345), (4764,))"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xo_test_rna.shape, xd_test.shape, y_test.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The shapes above match as expected. Specifically, all training data frames have the same number of drug cell line pairs. All testing data also has the same number of drug cell line pairs. The omics inputs have the same number of features. The durg inputs also have the same number of features."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build benchmark mean model\n",
    "Benchmark performance from just predicting the average for each drug in the training set.\n",
    "\n",
    "This does well because if  a drug is effective for a subset of cell lines it is normally good for most / all cell lines. Therefore, the mean values caputre this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.6957834043053239, 1.993983889326656, (0.8424264418165746, 0.0))"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mm = mean_model.MeanModel(y_train, _all_drugs)\n",
    "mm_pre = mm.predict(y_test.index)\n",
    "r2_score(y_test, mm_pre), mean_squared_error(y_test, mm_pre), pearsonr(y_test, mm_pre)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.1032089047619045"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#find the positions of drug Afatinib in y_train\n",
    "drugs = np.array([cl_d.split('::')[1] for cl_d in y_train.index]) \n",
    "drug_pos = np.where(drugs == 'Afatinib') #positions of Afatinib\n",
    "#average over all instances of 'Afatinib' in train set\n",
    "np.mean(y_train.iloc[drug_pos[0]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.1032089047619045"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mm.model['Afatinib'] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "above shows that mean model is simply a dict that maps a drug to the average target value of that drug in the training set. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model building  no hp tuning\n",
    "\n",
    "We now build a neural network model, using kears, for DRP. This network has two inputs, an omic input and a drug input. \n",
    "We will set all the hyperparameters for now. \n",
    "\n",
    "The omic input is passed through several 1D CNN layers, for feature extraction before being concatenated with the drug input. \n",
    "\n",
    "The output layer has no activation function because the target values are continuous "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_shape=None\n",
    "def build_cnn(hps=None):\n",
    "    if input_shape == None:\n",
    "        raise Exception('add input shape dim')\n",
    "    omic_input = layers.Input(shape=(input_shape, 1))\n",
    "    x = layers.Conv1D(\n",
    "        filters=16, kernel_size=16, activation='relu')(omic_input)\n",
    "    x = layers.MaxPooling1D(pool_size=2)(x)\n",
    "    x = layers.Conv1D(16, kernel_size=8, activation='relu')(x)\n",
    "    x = layers.MaxPooling1D(pool_size=2)(x)\n",
    "    x = layers.Flatten()(x)\n",
    "    x = layers.Dense(128, activation='relu')(x)\n",
    "    x = layers.Dense(128, activation='relu')(x)\n",
    "    drug_input = layers.Input(shape = (xd_train.shape[1]))\n",
    "    concatenated = layers.concatenate([x, drug_input])\n",
    "    hidd = layers.Dense(64, activation='relu')(concatenated)\n",
    "    hidd = layers.Dense(64, activation='relu')(hidd)\n",
    "    output_tensor = layers.Dense(1)(hidd)\n",
    "    model = tf.keras.Model([omic_input, drug_input], output_tensor)\n",
    "    opt = tf.keras.optimizers.RMSprop(learning_rate=1e-3)\n",
    "    model.compile(optimizer=opt, loss=tf.keras.metrics.mean_squared_error, metrics=['mae'])\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"model_1\"\n",
      "__________________________________________________________________________________________________\n",
      "Layer (type)                    Output Shape         Param #     Connected to                     \n",
      "==================================================================================================\n",
      "input_4 (InputLayer)            [(None, 908, 1)]     0                                            \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_4 (Conv1D)               (None, 893, 16)      272         input_4[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_4 (MaxPooling1D)  (None, 446, 16)      0           conv1d_4[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_5 (Conv1D)               (None, 439, 16)      2064        max_pooling1d_4[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_5 (MaxPooling1D)  (None, 219, 16)      0           conv1d_5[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "flatten_2 (Flatten)             (None, 3504)         0           max_pooling1d_5[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "dense_7 (Dense)                 (None, 128)          448640      flatten_2[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dense_8 (Dense)                 (None, 128)          16512       dense_7[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "input_5 (InputLayer)            [(None, 345)]        0                                            \n",
      "__________________________________________________________________________________________________\n",
      "concatenate_1 (Concatenate)     (None, 473)          0           dense_8[0][0]                    \n",
      "                                                                 input_5[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "dense_9 (Dense)                 (None, 64)           30336       concatenate_1[0][0]              \n",
      "__________________________________________________________________________________________________\n",
      "dense_10 (Dense)                (None, 64)           4160        dense_9[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "dense_11 (Dense)                (None, 1)            65          dense_10[0][0]                   \n",
      "==================================================================================================\n",
      "Total params: 502,049\n",
      "Trainable params: 502,049\n",
      "Non-trainable params: 0\n",
      "__________________________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "input_shape=xo_train_rna.shape[1]\n",
    "build_cnn().summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will use cross-validation to evaluate the model, using the run_cv_cblind method from my cross_val module. \n",
    "\n",
    "This implements cross-validation such that the cell lines in the validation set are not in the partial training set. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0% complete\n",
      "10.0% complete\n",
      "20.0% complete\n",
      "30.0% complete\n",
      "40.0% complete\n",
      "50.0% complete\n",
      "60.0% complete\n",
      "70.0% complete\n",
      "80.0% complete\n",
      "90.0% complete\n"
     ]
    }
   ],
   "source": [
    "loss, val_loss, val_mae, _ = cross_val.run_cv_cblind(\n",
    "    m_func=build_cnn, \n",
    "    x=[xo_train_rna, xd_train], \n",
    "    y=y_train,\n",
    "    hp=None,\n",
    "    epochs=1,\n",
    "    random_seed=50\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model building with hp tuning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tfGPUforge",
   "language": "python",
   "name": "tfgpuforge"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
